### PR TITLE
Optionally prepend react import if there is at least one TSX backend

### PR DESCRIPTION
**Pull request by Magellol**
Only prepend a react import once when there is at least one TSX backend

Again, this may not be idiomatic haskell 😅 

## TODO

- [ ] ~~Forgot to keep `ReactElement`~~

### CODE REVIEW

**Comment by Magellol** [comment]
> Forgot to keep ReactElement

@sam @OliverJAsh I'm guessing we could have translation file without any functions returning `ReactElement`? So we should also look at dropping it in that case?

File: lib/Intlc/Backend/TypeScript/Compiler.hs [hunk 1]
```diff
@@ -25,8 +25,14 @@ compileNamedExport x l k v =
...
-reactImport :: Text
-reactImport = "import React, { ReactElement } from 'react'"
+isTypeScriptReact :: Translation -> Bool
+isTypeScriptReact (Translation _m  be) = case be of
+  TypeScript      -> False
+  TypeScriptReact -> True
+
+
+buildReactImport :: Dataset Translation -> Maybe Text
+buildReactImport = fmap (const "import React, { ReactElement } from 'react'") . find isTypeScriptReact
```
Comment (samhh) [inline lib/Intlc/Backend/TypeScript/Compiler.hs:35]: I'd argue this belongs in the JS compiler, which handles the runtime stuff for TS.

File: lib/Intlc/Backend/TypeScript/Compiler.hs [hunk 2]
```diff
@@ -25,8 +25,14 @@ compileNamedExport x l k v =
 compileTypeof :: InterpStrat -> ICU.Message -> Text
 compileTypeof x = let o = fromStrat x in flip runReader o . typeof . fromMsg o
 
-reactImport :: Text
-reactImport = "import React, { ReactElement } from 'react'"
+isTypeScriptReact :: Translation -> Bool
+isTypeScriptReact (Translation _m  be) = case be of
+  TypeScript      -> False
+  TypeScriptReact -> True
```
Comment (samhh) [inline lib/Intlc/Backend/TypeScript/Compiler.hs:31]: Arguably a minor improvement would be to pattern match directly in the function parameter:

```hs
isTypeScriptReact (Translation _ TypeScriptReact) = True
isTypeScriptReact _ = False
```

Although that this function exists at all is another argument for lenses, or even just field selectors.

File: lib/Intlc/Backend/TypeScript/Compiler.hs (hunk 1 above)
Comment (samhh) [inline lib/Intlc/Backend/TypeScript/Compiler.hs:35]: We can move the string into a `where` clause to make the top-level algorithm more focused.

We're discarding the contents of the `Just` coming from `find`, so we could instead use `any`. In turn we'll need to change the code slightly to operate over `Bool`. There are a few different ways, here's one example:

```hs
flip pureIf txt . any isTypeScriptReact

-- or not pointfree, arguably more readable?
xs = pureIf (any isTypeScriptReact xs) txt
```

(where `txt` is in a `where` clause)

File: lib/Intlc/Compiler.hs [hunk 3]
```diff
@@ -13,11 +13,14 @@ import           Intlc.Core
 import qualified Intlc.ICU                         as ICU
 import           Prelude                           hiding (ByteString)
 
+prependOptionalReactImport :: Dataset Translation -> Text
+prependOptionalReactImport = maybe "" (<> "\n") . TS.buildReactImport
```
Comment (samhh) [inline lib/Intlc/Compiler.hs:17]: Because `Text` is a `Monoid`, we can replace `maybe ""` with `foldMap`.

**Review by samhh — APPROVED** [review]
Looks good! None of my feedback is a big deal. :slightly_smiling_face:

**Comment by samhh** [comment]
Regarding `ReactElement`, we only need it for the `tsx` backend, which I think this PR adheres to. The linked issue is to drop `React` once web no longer needs it, which'll just be a case of updating the import string.

File: lib/Intlc/Backend/TypeScript/Compiler.hs (hunk 2 above)
Comment (Magellol) [reply lib/Intlc/Backend/TypeScript/Compiler.hs:31]: Ah very cool!

**Pull request merged**